internal/impl/core/metrics.go

```diff
@@ -115,98 +119,111 @@ type ccsmpBackedMetrics struct {
+	if backedMetrics.capturedTxMetrics != nil {
 		return
 	}
-	metrics.capturedTxMetrics = make(map[ccsmp.SolClientStatsTX]uint64)
+	backedMetrics.capturedTxMetrics = make(map[ccsmp.SolClientStatsTX]uint64)
 	for _, txStat := range txMetrics {
-		metrics.capturedTxMetrics[txStat] = metrics.session.SolClientSessionGetTXStat(txStat)
+		backedMetrics.capturedTxMetrics[txStat] = backedMetrics.session.SolClientSessionGetTXStat(txStat)
 	}
 }
 
-func (metrics *ccsmpBackedMetrics) captureRXStats() {
-	if metrics.capturedRxMetrics != nil {
+func (backedMetrics *ccsmpBackedMetrics) captureRXStats() {
+	if backedMetrics.capturedRxMetrics != nil {
 		return
 	}
-	metrics.capturedRxMetrics = make(map[ccsmp.SolClientStatsRX]uint64)
+	backedMetrics.capturedRxMetrics = make(map[ccsmp.SolClientStatsRX]uint64)
 	for _, rxStat := range rxMetrics {
-		metrics.capturedRxMetrics[rxStat] = metrics.session.SolClientSessionGetRXStat(rxStat)
+		backedMetrics.capturedRxMetrics[rxStat] = backedMetrics.session.SolClientSessionGetRXStat(rxStat)
 	}
 }
 
-func (metrics *ccsmpBackedMetrics) getTXStat(stat ccsmp.SolClientStatsTX) uint64 {
-	metrics.metricLock.RLock()
-	defer metrics.metricLock.RUnlock()
-	if metrics.capturedTxMetrics != nil {
-		return metrics.capturedTxMetrics[stat]
+func (backedMetrics *ccsmpBackedMetrics) getTXStat(stat ccsmp.SolClientStatsTX) uint64 {
+	backedMetrics.metricLock.RLock()
+	defer backedMetrics.metricLock.RUnlock()
+	if backedMetrics.capturedTxMetrics != nil {
+		return backedMetrics.capturedTxMetrics[stat]
 	}
-	return metrics.session.SolClientSessionGetTXStat(stat)
+	return backedMetrics.session.SolClientSessionGetTXStat(stat)
 }
 
-func (metrics *ccsmpBackedMetrics) getRXStat(stat ccsmp.SolClientStatsRX) uint64 {
-	metrics.metricLock.RLock()
-	defer metrics.metricLock.RUnlock()
-	if metrics.capturedRxMetrics != nil {
-		return metrics.capturedRxMetrics[stat]
+func (backedMetrics *ccsmpBackedMetrics) getRXStat(stat ccsmp.SolClientStatsRX) uint64 {
+	backedMetrics.metricLock.RLock()
+	defer backedMetrics.metricLock.RUnlock()
+	if backedMetrics.capturedRxMetrics != nil {
+		return backedMetrics.capturedRxMetrics[stat]
 	}
-	return metrics.session.SolClientSessionGetRXStat(stat)
+	return backedMetrics.session.SolClientSessionGetRXStat(stat)
 }
 
-func (metrics *ccsmpBackedMetrics) getNextGenStat(metric NextGenMetric) uint64 {
-	return atomic.LoadUint64(&metrics.metrics[metric])
+func (backedMetrics *ccsmpBackedMetrics) getNextGenStat(metric NextGenMetric) uint64 {
+	return atomic.LoadUint64(&backedMetrics.metrics[metric])
 }
 
-func (metrics *ccsmpBackedMetrics) GetStat(metric metrics.Metric) uint64 {
+func (backedMetrics *ccsmpBackedMetrics) GetStat(metric metrics.Metric) uint64 {
 	if rxMetric, ok := rxMetrics[metric]; ok {
-		return metrics.getRXStat(rxMetric)
+		// check for duplicate counts due to auto-acks and remove from final result
+		rxStat := backedMetrics.getRXStat(rxMetric)
+		duplicateAck := uint64(0)
```

Comment:
Shouldn't this be using `backedMetrics.duplicateAcks` instead of this local var that is always 0?

Reply:
Thanks for pointing this out @sdewilde